### PR TITLE
Support for deployment into DDB Azure

### PR DESCRIPTION
**Pull request by cableman**
* Change image build
* Add helm chart

We might need help to get the actions up-to-date with these changes?

### CODE REVIEW

**Review by turegjorup — APPROVED** [review]

**Review by kasperg — COMMENTED** [review]
Remember to remove all remains of the existing infrastructure which will no longer work or are no longer relevant after this change. This includes:

- GitHub [actions](https://github.com/itk-dev/ddb-material-list/tree/feature/azure-infra/.github/actions/deployer) and [workflows to build, test and deploy images](https://github.com/itk-dev/ddb-material-list/blob/feature/azure-infra/.github/workflows/push.yml#L92-L146)
- [README content referring to Google Cloud](https://github.com/itk-dev/ddb-material-list/blob/feature/azure-infra/README.md#hosting)

File: README.md [hunk 1]
```diff
@@ -237,14 +236,77 @@ To get the names of requests (for use in hook file), use `dredd
...
+kind: Secret
+metadata:
+  namespace: {{ .Release.Namespace }}
+  name: {{ .Release.Name }}-basic-auth
+type: Opaque
+data:
+  auth: ''
+{{- end }}
+{{- end }}
+```
```
Comment (kasperg) [inline README.md]: I think it would be preferable to put this in a `secrets.example.yaml` file like we do for `.env`.

File: infrastructure/docker/material-list/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    && rm /tmp/app.tar
+
+# Add composer in from the official composer image (also alpine).
+COPY --from=composer:1.9.1 /usr/bin/composer /usr/bin/composer
+
+WORKDIR ${APP_PATH}
+
+## Install assets, which requires a HACK as redis is not available (should be removed later on).
+RUN composer install --no-dev --no-ansi --no-interaction --no-progress --no-suggest --prefer-dist --optimize-autoloader \
+    && composer clear-cache
```
Comment (kasperg) [inline infrastructure/docker/material-list/Dockerfile]: I do not understand the comment. The project does not use Redis.

File: infrastructure/docker/nginx/docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+set -eux
+
+## Run templates with configuration.
+/usr/local/bin/confd --onetime --backend env --confdir /etc/confd
+
+## Start the PHP process.
+/usr/sbin/nginx -g "daemon off;"
```
Comment (kasperg) [inline infrastructure/docker/nginx/docker-entrypoint.sh]:
```suggestion
## Start the nginx process.
/usr/sbin/nginx -g "daemon off;"
```

File: infrastructure/docker/nginx/etc/confd/templates/default.conf.tmpl [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    location @rewrite {
+        rewrite ^/(.*)$ /index.php?q=$1;
+    }
+
+    # Don't allow direct access to PHP files in the vendor directory.
+    location ~ /vendor/.*\.php$ {
+        deny all;
+        return 404;
+    }
```
Comment (kasperg) [inline infrastructure/docker/nginx/etc/confd/templates/default.conf.tmpl]: Is this relevant? In Laravel the vendor directory should not be exposed by the web server.

File: infrastructure/docker/nginx/etc/confd/templates/nginx.conf.tmpl [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+	sendfile on;
+	tcp_nopush on;
+	tcp_nodelay on;
+	keepalive_timeout 15 15;
+	types_hash_max_size 2048;
+
+	server_tokens off;
+
+    gzip  on;
```
Comment (kasperg) [inline infrastructure/docker/nginx/etc/confd/templates/nginx.conf.tmpl]: 🔬 Indentation is a bit off here.

File: infrastructure/material_list/LICENSE.txt [hunk 6]
```diff
@@ -0,0 +1,661 @@
+                    GNU AFFERO GENERAL PUBLIC LICENSE
+                       Version 3, 19 November 2007
```
Comment (kasperg) [inline infrastructure/material_list/LICENSE.txt]: Why include this? [We already have a license in project root](https://github.com/danskernesdigitalebibliotek/ddb-material-list/blob/master/LICENSE.md).

File: infrastructure/material_list/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v2
+name: material_list
+description: DDB Material list service
+
+type: application
+
+version: 0.1.0
+appVersion: 1.0.2
```
Comment (kasperg) [inline infrastructure/material_list/Chart.yaml]: What does `appVersion` refer to here? Is it possible to interpolate this from git tags instead of hardcoding it? 1.0.2 is not the current version of the application.

File: .github/workflows/pull_request.yml [hunk 8]
```diff
@@ -44,3 +44,15 @@ jobs:
...
         REVIEWDOG_TOKEN: ${{ secrets.REVIEWDOG_TOKEN }}
+
+  helm_validation:
+    name: Helm chart validation
+    runs-on: ubuntu-latest
+    steps:
+      - uses: azure/setup-helm@v1
+        name: Install helm
+        with:
+          version: v3.1.2
```
Comment (kasperg) [inline .github/workflows/pull_request.yml]: Is there a reason why we specify this exact version? It is not the latest stable version defined by `setup-helm` nor the latest version of Helm.

If there is a reason to use this version I think we should specify it in a comment or the like.

File: infrastructure/README.md [hunk 9]
```diff
@@ -1,37 +0,0 @@
-# Material-list hosting infrastructure
-This directory contains configuration, scripts and tools for hosting Material List. The current example is based on cloud hosting, but as it is based on  standard Open Source components it can be adapted to any cloud or traditional hosting.
-
-## Requirements
-Material-list requires a host with the following software:
-- Nginx 1.17
-- PHP 7.3
-- Mysql 5.7+ compatible database
-Individual components can be swapped for relevant alternatives. Apache can be used instead of Nginx. Any [database supported by Laravel](https://laravel.com/docs/5.8/database) such as PostgresSQL, SQLite and SQL Server can replace MariaDB.
```
Comment (kasperg) [inline infrastructure/README.md:9]: Perhaps it would be good to move this list of general requirements to the project `README.md`?

File: docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    networks:
+      - app
+    environment:
+      MYSQL_DATABASE: db
+      MYSQL_PASSWORD: db
+      MYSQL_ROOT_PASSWORD: root
+      MYSQL_USER: db
+
+  composer:
+    image: composer:1.9
```
Comment (kasperg) [inline docker-compose.yml]: This is not the most recent version of Composer 1. Is there a reason why we use this version and not 1 - or 1.10?

File: infrastructure/docker/material-list/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ARG APP_VERSION="develop"
+ENV APP_PATH=/var/www/html
+
+# Move site into the container.
+ADD https://github.com/danskernesdigitalebibliotek/ddb-material-list/archive/${APP_VERSION}.tar.gz /tmp/app.tar
+RUN tar -zxf /tmp/app.tar --strip-components=1 -C ${APP_PATH} \
+    && rm /tmp/app.tar
+
+# Add composer in from the official composer image (also alpine).
+COPY --from=composer:1.9.1 /usr/bin/composer /usr/bin/composer
```
Comment (kasperg) [inline infrastructure/docker/material-list/Dockerfile]: See previous comment regarding Composer version.

File: infrastructure/material_list/.helmignore [hunk 12]
```diff
@@ -0,0 +1,23 @@
+# Patterns to ignore when building packages.
+# This supports shell glob matching, relative path matching, and
+# negation (prefixed with !). Only one pattern per line.
+.DS_Store
+# Common VCS dirs
+.git/
+.gitignore
+.bzr/
```
Comment (kasperg) [inline infrastructure/material_list/.helmignore:8]: This includes many different options which does not seem relevant to the current project. Should they be removed?

File: infrastructure/material_list/values.yaml [hunk 13]
```diff
@@ -0,0 +1,49 @@
+# Default values for cover-service.
```
Comment (kasperg) [inline infrastructure/material_list/values.yaml]:
```suggestion
# Default values for material-list-service.
```

File: infrastructure/docker/php-fpm/config/90-materiallist.ini [hunk 14]
```diff
@@ -1,5 +0,0 @@
-; php.ini
-
-date.timezone = UTC
-expose_php = off
-memory_limit = 64M
```
Comment (kasperg) [inline infrastructure/docker/php-fpm/config/90-materiallist.ini:5]: This limit seems important. It does not seem to be replicated in the new setup. Is that intentional?

File: infrastructure/material_list/values.yaml [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  timezone: 'UTC'
+  cacheDriver: 'file'
+  queueConnection: 'sync'
+  cors_allow_origin: "*"
+  php:
+    max_children: "80"
+    start_servers: "15"
+    min_spare_servers: "15"
+    max_spare_servers: "30"
+    slowlog_timeout: "1"
```
Comment (kasperg) [inline infrastructure/material_list/values.yaml]: These values are a lot higher than the current values for production from the deleted `zz-pm.conf`. I do not know whether that is intentional or not.

File: config/database.php [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+            'charset' => env('DB_CHARSET', 'utf8mb4'),
+            'collation' => env('DB_COLLATION', 'utf8mb4_unicode_ci'),
+            'prefix' => env('DB_PREFIX', ''),
+            'strict' => env('DB_STRICT_MODE', true),
+            'engine' => env('DB_ENGINE', null),
+            'timezone' => env('DB_TIMEZONE', '+00:00'),
+            'options' => env('DB_SSL', false) ? array(
+                PDO::MYSQL_ATTR_SSL_KEY => base_path() . '/certs/DigiCertGlobalRootG2.crt.pem',
+                PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT => true,
+            ) : array(),
```
Comment (kasperg) [inline config/database.php]:
```suggestion
            'options' => env('DB_SSL', false) ? [
                PDO::MYSQL_ATTR_SSL_KEY => base_path() . '/certs/DigiCertGlobalRootG2.crt.pem',
                PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT => true,
            ] : [],
```

No reason to mix in old school arrays.

File: docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,55 @@
+version: "3"
+
+networks:
+  frontend:
+    external: true
+  app:
+    driver: bridge
+    internal: false
```
Comment (kasperg) [inline docker-compose.yml]: Is it intentional that this file should be usable in general or only as a part of the Azure infrastructure? I would expect the former since it is now placed in the root of the repository in but if I try to run `docker-compose up` I get the following error:

```
11:29 $ docker-compose up
WARNING: The COMPOSE_PROJECT_NAME variable is not set. Defaulting to a blank string.
WARNING: The COMPOSE_DOMAIN variable is not set. Defaulting to a blank string.
Creating network "material-list_app" with driver "bridge"
ERROR: Network frontend declared as external, but could not be found. Please create the network manually using `docker network create frontend` and try again.
```

If it is for general use and require special handling then please update the documentation accordingly. Otherwise I think it belongs in the infrastructure folder.

File: config/database.php [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+            'password' => env('DB_PASSWORD', ''),
+            'unix_socket' => env('DB_SOCKET', ''),
+            'charset' => env('DB_CHARSET', 'utf8mb4'),
+            'collation' => env('DB_COLLATION', 'utf8mb4_unicode_ci'),
+            'prefix' => env('DB_PREFIX', ''),
+            'strict' => env('DB_STRICT_MODE', true),
+            'engine' => env('DB_ENGINE', null),
+            'timezone' => env('DB_TIMEZONE', '+00:00'),
+            'options' => env('DB_SSL', false) ? array(
+                PDO::MYSQL_ATTR_SSL_KEY => base_path() . '/certs/DigiCertGlobalRootG2.crt.pem',
```
Comment (kasperg) [inline config/database.php:34]: Should this key be defined as a part of the codebase? I would have expected it to be an environment variable.

I see that it might cause problems because `MYSQL_ATTR_SSL_KEY` must be a file path and not a value but if it is to be part of the codebase it would be good to have some kind of documentation regarding the origin and maintenance of the file. There are no code comments and I could not spot anything in the commit message either.

**Review by kasperg — CHANGES_REQUESTED** [review]
:+1: I have some comments here and there. A lot of it may be caused by my own limited knowledge of Kubernetes. There is also some questions regarding our opinion on versioning of third party tools.

My main reason for blocking the PR in the current form is the addition of the root `docker-compose.yml` file which does not seem to work in the current form. Something is also off regarding our GitHub Actions. They have not been run for the latest changes in this PR. I do not know why.

File: README.md (hunk 1 above)
Comment (cableman) [reply README.md]: That would not work as helm picks up all yaml files.... I'll note it and take it into account in another PR if an solution exists from helm

File: infrastructure/material_list/LICENSE.txt (hunk 6 above)
Comment (cableman) [reply infrastructure/material_list/LICENSE.txt]: Default from helm to have licens file in all charts, but can remove it

File: docker-compose.yml (hunk 17 above)
Comment (cableman) [reply docker-compose.yml]: It not related to kubernetes it's was intended for development and was based on ITK docker composer setup. Removed it as it is not needed and ITK is not going to do development on this project.

File: README.md (hunk 1 above)
Comment (cableman) [reply README.md]: Will try using ignore in helm to add example file.

Comment (cableman) [reply README.md]: Fixed and moved into example file and readme updated.

File: infrastructure/docker/material-list/Dockerfile (hunk 2 above)
Comment (cableman) [reply infrastructure/docker/material-list/Dockerfile]: Comment updated

File: infrastructure/docker/nginx/etc/confd/templates/default.conf.tmpl (hunk 4 above)
Comment (cableman) [reply infrastructure/docker/nginx/etc/confd/templates/default.conf.tmpl]: No removed

File: infrastructure/docker/nginx/etc/confd/templates/nginx.conf.tmpl (hunk 5 above)
Comment (cableman) [reply infrastructure/docker/nginx/etc/confd/templates/nginx.conf.tmpl]: Fixed was hidden in my editor...

File: infrastructure/material_list/LICENSE.txt (hunk 6 above)
Comment (cableman) [reply infrastructure/material_list/LICENSE.txt]: Removed

File: infrastructure/material_list/Chart.yaml (hunk 7 above)
Comment (cableman) [reply infrastructure/material_list/Chart.yaml]: Noted and for now removed it's a version only visible to the administrator for the cluster using helm. An as we will use `set` command to set the container version is not needed.

File: .github/workflows/pull_request.yml (hunk 8 above)
Comment (cableman) [reply .github/workflows/pull_request.yml]: Updated to the match the version I currently use to install the charts. It needed as this will fallback to helm 2, which we don't use.

File: infrastructure/README.md (hunk 9 above)
Comment (cableman) [reply infrastructure/README.md:9]: Moved

File: docker-compose.yml (hunk 10 above)
Comment (cableman) [reply docker-compose.yml]: Changed to `1` :-)

File: infrastructure/docker/material-list/Dockerfile (hunk 11 above)
Comment (cableman) [reply infrastructure/docker/material-list/Dockerfile]: Changed to 1 :-)

File: infrastructure/material_list/.helmignore (hunk 12 above)
Comment (cableman) [reply infrastructure/material_list/.helmignore:8]: It's the default helm generates....

File: infrastructure/material_list/values.yaml (hunk 13 above)
Comment (cableman) [reply infrastructure/material_list/values.yaml]: Fixed

File: config/database.php (hunk 18 above)
Comment (cableman) [reply config/database.php:34]: It's MS Azure managed root certificate and as this is outside the k8s network it needs to be SSL enabled (it's actually forced by MS).

And comment and link about this file.

File: infrastructure/material_list/values.yaml (hunk 15 above)
Comment (cableman) [reply infrastructure/material_list/values.yaml]: The values was based on requests to cover service... adjusted them to match the old values, but will be re-adjusted when we get load on the setup.

File: infrastructure/docker/php-fpm/config/90-materiallist.ini (hunk 14 above)
Comment (cableman) [reply infrastructure/docker/php-fpm/config/90-materiallist.ini:5]: Yes and changed to default to 64M. The other vaules are default.

**Comment by cableman** [comment]
In regards to why actions are not running: There are something about PR's from a fork form a user thats not owner (me) don't run if there are changes to the workflows to protect that actions can't be mis-used.

**Comment by cableman** [comment]
Changed the different things from review and re-deploy to test setup.

File: .github/workflows/pull_request.yml (hunk 8 above)
Comment (kasperg) [reply .github/workflows/pull_request.yml]: How do you see it fall back to version 2? As I see it [the action tries to retrieve a version greater than 3.2.1 from GitHub](https://github.com/Azure/setup-helm/blob/master/src/run.ts#L41).

**Review by kasperg — APPROVED** [review]
:+1: I still have some quibbles regarding the Helm version. Otherwise LGTM. Thanks for the updates.
